### ISSUE TITLE
View Applications needs rewriting (Distrobox 1.7)

### ISSUE THREAD

**Issue by Dvlv** [Dvlv/BoxBuddyRS#72]
https://cubiclenate.com/2024/03/02/distrobox-with-boxbuddy-on-opensuse/

**Comment by VortexAcherontic** [comment]
This once worked. I am not sure if an update in Tumbleweed broke it, distrobox or boxbuddy. Did not yet wrap my head around it 😄 
Happens with podman as container runtime as well.  
I'll try to figure out what actually is broken.

**Comment by VortexAcherontic** [comment]
Just found another issue which is totally up to distrobox. Since it's latest release and their "improvements" to the nvidia driver integration distrobox will get stuck at "Setting up host's nvidia integration..." and never proceed. You have to enter the container a 2nd time until it works.

As fetching installed apps is done using `distrobox enter tumbleweed -- bash -c grep -L "NoDisplay=true" /usr/share/applications/*.desktop` this feature is broken as of distrobox 1.7.0 on nvidia distroboxes.

**Comment by VortexAcherontic** [comment]
Regarding the actuall issue. My findgns so far are two blank lines in the command output of `distrobox enter tumbleweed -- bash -c grep -L "NoDisplay=true" /usr/share/applications/*.desktop`

Example output from my rocky linux box:

```
>  GSETTINGS_SCHEMA_DIR=../../ ./boxbuddy-rs 
/usr/share/applications/com.blackmagicdesign.rawplayer.desktop
/usr/share/applications/com.blackmagicdesign.rawspeedtest.desktop
/usr/share/applications/com.blackmagicdesign.resolve-CaptureLogs.desktop
/usr/share/applications/com.blackmagicdesign.resolve-Installer.desktop
/usr/share/applications/com.blackmagicdesign.resolve-Panels.desktop
/usr/share/applications/com.blackmagicdesign.resolve.desktop
/usr/share/applications/nvidia-settings.desktop


```

What is interesting though is, it actually returns the list of applications but seems not to properly parse them for display in the UI.

**Comment by VortexAcherontic** [comment]
Hm, looking further into the matter I found the follwoing:  

```
/usr/bin/distrobox-enter: eval: line 614: unexpected EOF while looking for matching `"'


/usr/bin/distrobox-enter: eval: line 614: unexpected EOF while looking for matching `"'


/usr/bin/distrobox-enter: eval: line 614: unexpected EOF while looking for matching `"'


/usr/bin/distrobox-enter: eval: line 614: unexpected EOF while looking for matching `"'


/usr/bin/distrobox-enter: eval: line 614: unexpected EOF while looking for matching `"'


/usr/bin/distrobox-enter: eval: line 614: unexpected EOF while looking for matching `"'


/usr/bin/distrobox-enter: eval: line 614: unexpected EOF while looking for matching `"'
```

Which is what get_pieces_cmd actually contains after:  
```
let get_pieces_cmd = get_command_output(
            String::from("distrobox"),
            Some(&[
                "enter",
                &box_name,
                "--",
                "bash",
                "-c",
                &format!(
                    "NAME=$(grep -m 1 \"^Name=\" {} 
            | sed 's/^Name=//'); 
            EXEC=$(grep -m 1 \"^Exec=\" {} 
            | sed 's/^Exec=//' | tr -d '\n'); 
            ICON=$(grep -m 1 \"^Icon=\" {} 
            | sed 's/^Icon=//' | tr -d '\n'); 
            echo \"${{NAME}} | ${{EXEC}} | ${{ICON}}\"",
                    line, line, line
                ),
            ]),
        );
```

**Comment by Dvlv** [comment]
Yeah, I traced it back to that same place. Updates to Distrobox have changed how the shelling-out is handled, so now the stuff in `$()` is being executed on the host, when it was previously being done in the container.

This means if I have `geany` in the container but not on the host, it's doing:
```
$(grep -m 1 \"^Name=\" /usr/share/applications/geany.desktop | sed 's/^Name=//')
```
on the host, which has no `/usr/share/geany.desktop`.

I'm gonna have to redo this to do all the parsing in Rust, which is a pain. When I was doing it in python it was unusably slow; of course Rust is faster than Python but I'll need to see if it's acceptable.

I assume shelling-out to distrobox is fairly slow, so I'm wondering if I can be clever and do something like

```
for df in desktopFiles do
    cat df
    echo "---BoxBuddy---"
end
```
(not real bash syntax)

Then get one big string, which I can split on the `---BoxBuddy---` into a Vec, so I'm only shelling out once.

**Comment by Dvlv** [comment]
Fix pushed to master

**Comment by VortexAcherontic** [comment]
You are a true magician. It works very well. Many thanks. ❤️

**Comment by Dvlv** [comment]
2.1.4 now on flathub
